src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -389,6 +406,12 @@ private async Task StopAsyncCore(bool disposing)
             {
                 await connectionState.StopAsync();
             }
+
+            if (_invocationMessageChannel != null)
+            {
+                _invocationMessageChannel.Writer.TryComplete();
+                await _invocationMessageReceiveTask;
```

Comment:
Move this earlier into the method inside the connection lock.

Seeing the `await connectionState.StopAsync()` outside the connection lock above, I went down the rabbit hole of learning about the private ConnectionState class. I'm now wondering when we stopped using _connectionState to ["// Represent all the transient state about a connection"](https://github.com/aspnet/AspNetCore/blob/ee5d6728879a71eb4b2b16cd6786d28711815563/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L1232)? It was purpose-built for stuff like _invocationMessageChannel and _invocationMessageReceiveTask. I was about to recommend moving the fields, but the are now so many *other* fields defined directly on the HubConnection class that track transient state to the connection, it hardly seems worthwhile.

I [filed an issue](https://github.com/aspnet/AspNetCore/issues/8500) to address that and to make sure that all the references to _connectionState are even safe to begin with. I backlogged it, because I can't really justify spending time right now on a massive refactoring of HubConnection. It's too bad, because using ConnectionState correctly could help us avoid bugs. Perhaps if we find a bunch of HubConnection bugs cropping up, we'll reprioritize.

Reply:
No thinking about that, I think we should move this to the connection state, or at least into the receive loop itself.

Reply:
@mikaelm12 lets change this to a local function and local variables. There should be no state inside of the HubConnection, this logic should live and die as part of the ReceiveLoop call.

Reply:
Okay, got it 👍